### PR TITLE
DATAUP-636 xSV happy path setup

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes

The first step for setting up the xSV stuff is, in my opinion, factoring out the code that sets up data import app cells (both standard and bulk import). This PR does that, creating a new `importSetup.js` module. 

This is really just a refactor at this point with stubs for the flow of how the xSV integration will work. It takes the separate `initBulkImport` and `initImportApp` functions in the `stagingAreaViewer` module and combines them into a single `initImport` function. This then calls out to the new module that does all the logic around what file gets uploaded in which cell and how.

# Jira Ticket / Issue #

Related Jira ticket: https://kbase-jira.atlassian.net/browse/DATAUP-636
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
* there should be no visual change
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook
- [x] Any dependent changes have been merged and published in downstream modules

# Updating Version and Release Notes (if applicable)

- [n/a] [Version has been bumped](https://semver.org/) for each release
- [n/a] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

### CODE REVIEW

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 1]
```diff
@@ -179,7 +179,9 @@ define([
 
             // Set up the link to the web upload app.
             this.$elem.find('.web_upload_div').click(() => {
-                this.initImportApp('web_upload');
+                return Import.setupWebUploadCell().then(() => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:182]: Wrapped a version of the setup cell distinctly for the web uploader app, so it doesn't need to do anything magical with inputs.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 2]
```diff
@@ -1103,15 +1105,13 @@ define([
         },
 
         enableImportButton: function () {
-            const stagingAreaViewer = this;
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:1106]: not necessary with the arrow function.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 3]
```diff
@@ -1128,90 +1128,28 @@ define([
          *
          * Creating a new bulk import cell returns a Promise, so this returns a Promise.
          */
-        initBulkImport: function () {
+        initImport: function () {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:1131]: Simplified `initBulkImport` and `initImportApp` into this function, which just compiles all selected files with their type and passes them off to be imported.

File: test/unit/spec/narrative_core/upload/importSetup-spec.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            const typeCombos = [
+                ['FASTQ', 'SRA'],
+                ['FASTQ', 'MATRIX'],
+                ['FASTQ', 'SRA', 'MATRIX'],
+                ['FASTQ', 'MATRIX', 'MEDIA'],
+                ['FASTQ', 'SRA', 'MATRIX', 'MEDIA'],
+            ].concat(Object.keys(TYPE).map((singleType) => [singleType]));
+            fileCounts.forEach((numFiles) => {
+                typeCombos.forEach((typeCombo) => {
+                    const types = typeCombo.map((type) => TYPE[type]);
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/importSetup-spec.js:109]: I may have gotten a little too carried away with being clever here.

The main tests that happen are adding lots of files and type permutations, sending those to `setupImportCells` and tracking how the various cells get created. This includes some combinations of types, along with each individual type on its own, with a number of files ranging 1, 2, 5 of each type. It then combines each of those to make unique files for all types, runs the function, and sees that the cells get created with expected values (bulk import cells only use configured bulk apps, singleton import cells only use the right files, etc.)

There's probably more that could be done here, but that will come with better error coverage. Right now, this is all happy path-ish.

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 5]
```diff
@@ -260,7 +260,7 @@ define([
             const placeholder = $container.find('span.select2-selection__placeholder').html();
             expect(placeholder).toContain('Select a type');
 
-            //The options that should be in the import as dropdown
+            // The options that should be in the import as dropdown
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:263]: This has tweaks to keep up with the function changes. Also adds a test for the `Upload with URL` button.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/importSetup.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+) => {
+    'use strict';
+    const uploaders = Config.get('uploaders');
+    const bulkIds = new Set(uploaders.bulk_import_types);
+
+    function getSpreadsheetFileInfo(fileName) {
+        return Promise.resolve(fileName);
+    }
+
+    function initSingleFileUploads(fileInfo) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/importSetup.js:16]: Copied almost without change from the `stagingAreaViewer` code.

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 7]
```diff
@@ -283,7 +283,7 @@ define([
         it('renders the dropdown correctly when a type is selected', async () => {
             await stagingViewer.render();
 
-            //find the fake sra reads row specifically (via the download button, then chaining back up to the select dropdown above - since we don't have a unique ID for these select drodpowns it's the best mehtod for now)
+            // find the fake sra reads row specifically (via the download button, then chaining back up to the select dropdown above - since we don't have a unique ID for these select drodpowns it's the best mehtod for now)
```
Comment (ialarmedalien) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:286]: typo in `method`

**Review by ialarmedalien — APPROVED** [review]
Minor typo to fix, but otherwise looks fine! I will leave it to your discretion whether you want to fix the typo now or in a future edit.

**Comment by briehl** [comment]
> Minor typo to fix, but otherwise looks fine! I will leave it to your discretion whether you want to fix the typo now or in a future edit.

I'll get it in the next one. Thanks!

**Pull request merged**